File: src/network/libqhostinfo.cpp
#include <QHostAddress>
#include <QHostInfo>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qhostinfo.h>
#include "libqhostinfo.h"
#include "libqhostinfo.hxx"

QHostInfo* QHostInfo_new() {
    return new QHostInfo();
}

QHostInfo* QHostInfo_new2(const QHostInfo* d) {
    return new QHostInfo(*d);
}

QHostInfo* QHostInfo_new3(int lookupId) {
    return new QHostInfo(static_cast<int>(lookupId));
}

void QHostInfo_OperatorAssign(QHostInfo* self, const QHostInfo* d) {
    self->operator=(*d);
}

void QHostInfo_Swap(QHostInfo* self, QHostInfo* other) {
    self->swap(*other);
}

libqt_string QHostInfo_HostName(const QHostInfo* self) {
    QString _ret = self->hostName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QHostInfo_SetHostName(QHostInfo* self, const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    self->setHostName(name_QString);
}

libqt_list /* of QHostAddress* */ QHostInfo_Addresses(const QHostInfo* self) {
    QList<QHostAddress> _ret = self->addresses();
    // Convert QList<> from C++ memory to manually-managed C memory
    QHostAddress** _arr = static_cast<QHostAddress**>(malloc(sizeof(QHostAddress*) * _ret.size()));
    for (size_t i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QHostAddress(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void QHostInfo_SetAddresses(QHostInfo* self, const libqt_list /* of QHostAddress* */ addresses) {
    QList<QHostAddress> addresses_QList;
    addresses_QList.reserve(addresses.len);
    QHostAddress** addresses_arr = static_cast<QHostAddress**>(addresses.data);
    for (size_t i = 0; i < addresses.len; ++i) {
        addresses_QList.push_back(*(addresses_arr[i]));
    }
    self->setAddresses(addresses_QList);
}

int QHostInfo_Error(const QHostInfo* self) {
    return static_cast<int>(self->error());
}

void QHostInfo_SetError(QHostInfo* self, int errorVal) {
    self->setError(static_cast<QHostInfo::HostInfoError>(errorVal));
}

libqt_string QHostInfo_ErrorString(const QHostInfo* self) {
    QString _ret = self->errorString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QHostInfo_SetErrorString(QHostInfo* self, const libqt_string errorString) {
    QString errorString_QString = QString::fromUtf8(errorString.data, errorString.len);
    self->setErrorString(errorString_QString);
}

void QHostInfo_SetLookupId(QHostInfo* self, int id) {
    self->setLookupId(static_cast<int>(id));
}

int QHostInfo_LookupId(const QHostInfo* self) {
    return self->lookupId();
}

void QHostInfo_AbortHostLookup(int lookupId) {
    QHostInfo::abortHostLookup(static_cast<int>(lookupId));
}

QHostInfo* QHostInfo_FromName(const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return new QHostInfo(QHostInfo::fromName(name_QString));
}

libqt_string QHostInfo_LocalHostName() {
    QString _ret = QHostInfo::localHostName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string QHostInfo_LocalDomainName() {
    QString _ret = QHostInfo::localDomainName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QHostInfo_Delete(QHostInfo* self) {
    delete self;
}

File: src/network/libqhostinfo.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qhostinfo_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qhostinfo.html
pub const qhostinfo = struct {
    /// New constructs a new QHostInfo object.
    ///
    ///
    pub fn New() QtC.QHostInfo {
        return qtc.QHostInfo_new();
    }

    /// New2 constructs a new QHostInfo object.
    ///
    /// ``` d: QtC.QHostInfo ```
    pub fn New2(d: ?*anyopaque) QtC.QHostInfo {
        return qtc.QHostInfo_new2(@ptrCast(d));
    }

    /// New3 constructs a new QHostInfo object.
    ///
    /// ``` lookupId: i32 ```
    pub fn New3(lookupId: i32) QtC.QHostInfo {
        return qtc.QHostInfo_new3(@intCast(lookupId));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#operator-eq)
    ///
    /// ``` self: QtC.QHostInfo, d: QtC.QHostInfo ```
    pub fn OperatorAssign(self: ?*anyopaque, d: ?*anyopaque) void {
        qtc.QHostInfo_OperatorAssign(@ptrCast(self), @ptrCast(d));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#swap)
    ///
    /// ``` self: QtC.QHostInfo, other: QtC.QHostInfo ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHostInfo_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#hostName)
    ///
    /// ``` self: QtC.QHostInfo, allocator: std.mem.Allocator ```
    pub fn HostName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_HostName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.HostName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#setHostName)
    ///
    /// ``` self: QtC.QHostInfo, name: []const u8 ```
    pub fn SetHostName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.struct_libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        qtc.QHostInfo_SetHostName(@ptrCast(self), name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#addresses)
    ///
    /// ``` self: QtC.QHostInfo, allocator: std.mem.Allocator ```
    pub fn Addresses(self: ?*anyopaque, allocator: std.mem.Allocator) []QtC.QHostAddress {
        const _arr: qtc.struct_libqt_list = qtc.QHostInfo_Addresses(@ptrCast(self));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.QHostAddress, _arr.len) catch @panic("qhostinfo.Addresses: Memory allocation failed");
        const _data: [*]QtC.QHostAddress = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#setAddresses)
    ///
    /// ``` self: QtC.QHostInfo, addresses: []QtC.QHostAddress ```
    pub fn SetAddresses(self: ?*anyopaque, addresses: []QtC.QHostAddress) void {
        const addresses_list = qtc.struct_libqt_list{
            .len = addresses.len,
            .data = @ptrCast(addresses.ptr),
        };
        qtc.QHostInfo_SetAddresses(@ptrCast(self), addresses_list);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#error)
    ///
    /// ``` self: QtC.QHostInfo ```
    pub fn Error(self: ?*anyopaque) i64 {
        return qtc.QHostInfo_Error(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#setError)
    ///
    /// ``` self: QtC.QHostInfo, errorVal: qhostinfo_enums.HostInfoError ```
    pub fn SetError(self: ?*anyopaque, errorVal: i64) void {
        qtc.QHostInfo_SetError(@ptrCast(self), @intCast(errorVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#errorString)
    ///
    /// ``` self: QtC.QHostInfo, allocator: std.mem.Allocator ```
    pub fn ErrorString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_ErrorString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.ErrorString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#setErrorString)
    ///
    /// ``` self: QtC.QHostInfo, errorString: []const u8 ```
    pub fn SetErrorString(self: ?*anyopaque, errorString: []const u8) void {
        const errorString_str = qtc.struct_libqt_string{
            .len = errorString.len,
            .data = errorString.ptr,
        };
        qtc.QHostInfo_SetErrorString(@ptrCast(self), errorString_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#setLookupId)
    ///
    /// ``` self: QtC.QHostInfo, id: i32 ```
    pub fn SetLookupId(self: ?*anyopaque, id: i32) void {
        qtc.QHostInfo_SetLookupId(@ptrCast(self), @intCast(id));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#lookupId)
    ///
    /// ``` self: QtC.QHostInfo ```
    pub fn LookupId(self: ?*anyopaque) i32 {
        return qtc.QHostInfo_LookupId(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#abortHostLookup)
    ///
    /// ``` lookupId: i32 ```
    pub fn AbortHostLookup(lookupId: i32) void {
        qtc.QHostInfo_AbortHostLookup(@intCast(lookupId));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#fromName)
    ///
    /// ``` name: []const u8 ```
    pub fn FromName(name: []const u8) QtC.QHostInfo {
        const name_str = qtc.struct_libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.QHostInfo_FromName(name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#localHostName)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn LocalHostName(allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_LocalHostName();
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.LocalHostName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#localDomainName)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn LocalDomainName(allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_LocalDomainName();
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.LocalDomainName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhostinfo.html#dtor.QHostInfo)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QHostInfo ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QHostInfo_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qhostinfo.html#types
pub const enums = struct {
    pub const HostInfoError = enum {
        pub const NoError: i32 = 0;
        pub const HostNotFound: i32 = 1;
        pub const UnknownError: i32 = 2;
    };
};
